fdo.go

```diff
@@ -0,0 +1,264 @@
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		return ""
+	}
+	for _, extension := range iconExtensions {
+		//Example is /usr/share/icons/icon_theme/32x32/apps/xterm.png
+		testIcon := filepath.Join(dir, string(fyconSize), "apps", iconName+extension)
+		if _, err := os.Stat(testIcon); os.IsNotExist(err) {
+			testIcon := filepath.Join(dir, string(fyconSize)+"x"+string(fyconSize), "apps", iconName+extension)
+			if _, err := os.Stat(testIcon); os.IsNotExist(err) {
+				testIcon := filepath.Join(dir, "apps", string(fyconSize), iconName+extension)
+				if _, err := os.Stat(testIcon); os.IsNotExist(err) {
+					testIcon := filepath.Join(dir, "apps", string(fyconSize)+"x"+string(fyconSize), iconName+extension)
+					if _, err := os.Stat(testIcon); os.IsNotExist(err) {
+						continue
+					} else {
+						return testIcon
+					}
+				} else {
+					return testIcon
+				}
+			} else {
+				return testIcon
+			}
+		} else {
+			return testIcon
+		}
+	}
+	for _, extension := range iconExtensions {
+		//Example is /usr/share/icons/icon_theme/scalable/apps/xterm.png - Try this if the requested size didn't exist
+		testIcon := filepath.Join(dir, "scalable", "apps", iconName+extension)
+		if _, err := os.Stat(testIcon); os.IsNotExist(err) {
+			testIcon := filepath.Join(dir, "apps", "scalable", iconName+extension)
+			if _, err := os.Stat(testIcon); os.IsNotExist(err) {
+				continue
+			} else {
+				return testIcon
+			}
+		} else {
+			return testIcon
+		}
+	}
+	//If the requested icon wasn't found in the Example 32x32 or scalable dirs of the theme, try other sizes within theme
+	testIcon := reverseWalkDirectoryMatch(dir, "apps", iconName)
+	if testIcon != "" {
+		return testIcon
+	}
+
+	//One last chance at finding it - assume apps comes before size in path name
+	testIcon = reverseWalkDirectoryMatch(filepath.Join(dir, "apps"), "", iconName)
+
+	return testIcon
+}
+
+//FdoLookupIconPath will take the name of an Icon and find a matching image file
+func FdoLookupIconPath(iconName string) string {
+	locationLookup := FdoLookupXdgDataDirs()
+	for _, dataDir := range locationLookup {
+		//Example is /usr/share/icons/icon_theme
+		dir := filepath.Join(dataDir, "icons", fyconTheme)
+		iconPath := lookupIconPathInTheme(dir, iconName)
+		if iconPath != "" {
+			return iconPath
+		}
+	}
+	for _, dataDir := range locationLookup {
+		//Hicolor is the default fallback theme - Example /usr/share/icons/icon_theme/hicolor
+		dir := filepath.Join(dataDir, "icons", "hicolor")
+		iconPath := lookupIconPathInTheme(dir, iconName)
+		if iconPath != "" {
+			return iconPath
+		}
+	}
+	for _, dataDir := range locationLookup {
+		//Icon was not found in default theme or default fallback theme - Check all themes for a match
+		//Example is /usr/share/icons
+		files, err := ioutil.ReadDir(dataDir + "icons")
+		if err != nil {
+			fyne.LogError("", err)
+			continue
+		}
+		//Enter icon theme
+		for _, f := range files {
```

Comment:
This whole block gets very complex. I think inverting the logic, as above, should help simplify it.
Less indentation is a good indication of lower complexity.